src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy

```diff
@@ -81,129 +89,115 @@ class QinQFlowSpec extends HealthCheckSpecification {
         def traffExam = traffExamProvider.get()
         def examQinQFlow
         if(!trafficDisclaimer) {
-            examQinQFlow = new FlowTrafficExamBuilder(topology, traffExam)
-                    .buildBidirectionalExam(flowHelperV2.toV1(qinqFlow), 1000, 5)
-            withPool {
-                [examQinQFlow.forward, examQinQFlow.reverse].eachParallel { direction ->
-                    def resources = traffExam.startExam(direction)
-                    direction.setResources(resources)
-                    assert traffExam.waitExam(direction).hasTraffic()
-                }
-            }
+            examQinQFlow = qinqFlow.traffExam(traffExam, 1000, 5)
+            verifyFlowHasBidirectionalTraffic(examQinQFlow, traffExam)
         }
 
         and: "Involved switches pass switch validation"
-        def involvedSwitchesFlow1 = pathHelper.getInvolvedSwitches(
-                pathHelper.convert(northbound.getFlowPath(qinqFlow.flowId))
-        )
-        switchHelper.synchronizeAndCollectFixedDiscrepancies(involvedSwitchesFlow1*.getDpId()).isEmpty()
+        def involvedSwitchesFlow1 = qinqFlow.retrieveAllEntityPaths().getInvolvedSwitches()
+        switchHelper.synchronizeAndCollectFixedDiscrepancies(involvedSwitchesFlow1).isEmpty()
 
         when: "Create a vlan flow on the same port as QinQ flow"
-        def vlanFlow = flowHelper.randomFlow(swPair).tap {
-            it.source.portNumber = qinqFlow.source.portNumber
-            it.source.vlanId = qinqFlow.source.vlanId + 1
-            it.destination.portNumber = qinqFlow.destination.portNumber
-            it.destination.vlanId = qinqFlow.destination.vlanId + 1
-        }
-        flowHelper.addFlow(vlanFlow)
+        def vlanFlow = flowFactory.getBuilder(swPair)
+                .withSourcePort(qinqFlow.source.portNumber)
+                .withSourceVlan(qinqFlow.source.vlanId + 1)
+                .withDestinationPort(qinqFlow.destination.portNumber)
+                .withDestinationVlan(qinqFlow.destination.vlanId + 1)
+                .build().create()
 
         then: "Both existing flows are valid"
-        [qinqFlow.flowId, vlanFlow.id].each {
-            northbound.validateFlow(it).each { assert it.asExpected }
+        [qinqFlow, vlanFlow].each {
+            it.validateAndCollectDiscrepancies().isEmpty()
         }
 
         and: "Involved switches pass switch validation"
-        def involvedSwitchesFlow2 = pathHelper.getInvolvedSwitches(pathHelper.convert(northbound.getFlowPath(vlanFlow.id)))
-        def involvedSwitchesforBothFlows = (involvedSwitchesFlow1 + involvedSwitchesFlow2).unique { it.dpId }
-        switchHelper.synchronizeAndCollectFixedDiscrepancies(involvedSwitchesforBothFlows*.getDpId()).isEmpty()
+        def involvedSwitchesFlow2 = vlanFlow.retrieveAllEntityPaths().getInvolvedSwitches()
+        def involvedSwitchesforBothFlows = (involvedSwitchesFlow1 + involvedSwitchesFlow2).unique { it.id }
```

Comment:
I am not sure that there is `id `param, but you can use .unique()

Reply:
Done. Yes, there was `id` field there, however .unique() is simplier